### ISSUE TITLE
Exceptions during parsing

### ISSUE THREAD

**Issue by talex5** [mirage/mirage-tcpip#305]
I get several crashes when running this AFL fuzz test on mirage-nat, which exercises the parsing of ethernet frames, IPv4 packets, TCP, UDP and ICMPv4:

```
let read_input () =
  let b = Buffer.create 1000 in
  let buf = Bytes.create 1000 in
  try
    while true do
      let got = input stdin buf 0 (Bytes.length buf) in
      if got = 0 then raise End_of_file;
      assert (got > 0);
      Buffer.add_subbytes b buf 0 got
    done;
    assert false
  with End_of_file ->
    Buffer.contents b

let f () =
  let frame = Cstruct.of_string (read_input ()) in
  match Nat_packet.of_ethernet_frame frame with
  | Error _ -> ()
  | Ok p ->
    let packet = Cstruct.concat (Nat_packet.to_cstruct p) in
    match Nat_packet.of_ipv4_packet packet with
    | Error e -> failwith (Fmt.strf "Can't parse packet we created: %a" Nat_packet.pp_error e)
    | Ok p2 -> assert (Nat_packet.equal p p2)

let () =
  AflPersistent.run f
```

I'm testing tcpip 3.1.0 with OCaml 4.04.0+afl (from https://github.com/stedolan/opam-repository) and https://github.com/yomimono/ocaml-afl-persistent.

I'll list the crashes as comments on this issue.

**Comment by talex5** [comment]
```
0000000 5c68 4058 0200 0000 df00 00ff 0008 e84c
0000010 5703 585c 2b2b 2b2b 2b2b 2b2b 2b2b 2b2b
0000020 2b2b 2b2b 2b2b 2b2b 2b2b f52b 00f5 ff80
0000030 f1ff 
```
Gives:
```
Fatal error: exception (Invalid_argument "Cstruct.sub: [14,36](50) off=20 len=28")
Raised at file "pervasives.ml", line 33, characters 20-45
Called from file "lib/ipv4/ipv4_packet.ml", line 125, characters 42-99
Called from file "lib/nat_packet.ml", line 39, characters 8-47
Called from file "fuzz.ml", line 17, characters 8-42
```
Looks like we don't check that `hlen` is in range.

**Comment by talex5** [comment]
```
0000000 5ce8 5758 3e1f 00d5 ff48 00ff 0008 e847
0000010 4700 f5e8 e610 0100 f5ef 000f ef01 0ff5
0000020 d53e 4800 d53e 5959 0059 5904 5959 5959
0000030 5959 5948 2000 5959 5959 5959 5959 5959
0000040 5959 4800 ffe2 0800 f5f5 f56f f5f5 f5f5
0000050 0ff6 dce5 8800 ffad  
```
Gives:
```
Fatal error: exception (Failure
   "Can't parse packet we created: Failed to parse ICMP error's payload: Payload buffer (0 bytes) too small to contain payload (of size 35 from header)\
  \n48 00 00 43 00 00 00 00 59 59 80 79 59 59 59 59 \
  \n00 48 e2 ff 00 08 f5 f5 6f f5 f5 f5 f5 f5 f6 0f \
  \n")
```

**Comment by talex5** [comment]
```
0000000 5ce8 5758 3e1f 00d5 ff48 00ff 0008 e847
0000010 4700 f5e8 f610 1180 f5f5 ffff 7fff ffff
0000020 0800 4700 00e8 e847 10f5 6ff6 f5f5 3000
0000030 0800 4700 00e8 e847 10f5 80f6 f5f5 fff5
0000040 ffff ff7f 00ff 0008 e847 f500 f56f f5f5
0000050 f5f5 f5f5 f5f5 00f5
```
Gives:
```
Fatal error: exception (Invalid_argument "Cstruct.sub: [42,43](88) off=8 len=40")
Raised at file "pervasives.ml", line 33, characters 20-45
Called from file "lib/udp/udp_packet.ml", line 40, characters 18-65
Called from file "lib/nat_packet.ml", line 51, characters 18-59
```
We check if `payload_len > length_of_buffer`, but that's the total buffer length, not the payload area's length I think.

**Comment by talex5** [comment]
```
0000000 5ce8 5774 3e1f 00d5 ff48 00ff 0008 7f45
0000010 4700 f5e8 f610 0680 f5f5 ffff 7fff ffff
0000020 0800 a400 a4a4 a4a4 a4a4 a4a4 a4a4 a4a4
0000030 47a4 00e8 e847 1305 4700 00e8 e847 10f5
0000040 80f6 f5f5 fff5 ffff ff7f 00ff 0008 e847
0000050 f5f5 f5dd 00f5    
```
says the packet changed after writing and reloading:
```
IPv4 packet 255.255.255.127 -> 255.255.0.8: proto 6, ttl 128, options  TCP packet seq=2762253476 acknum=2762253476 ack=false rst=true syn=false fin=false win=42055 options=
[SACK[[4135646709,4127195135][4712448,1206449424]]] with payload ff 00 08 00 47 e8 f5 f5 dd f5 f5 
After:
IPv4 packet 255.255.255.127 -> 255.255.0.8: proto 6, ttl 128, options  TCP packet seq=2762253476 acknum=2762253476 ack=false rst=true syn=false fin=false win=42055 options=
[SACK[[4712448,1206449424][4135646709,4127195135]]] with payload ff 00 08 00 47 e8 f5 f5 dd f5 f5
```
Not sure if that matters, but seems odd.

**Comment by hannesm** [comment]
will close this issue. @yomimono fixed the issues in #307. please reopen if there's still something to do here.
